src/io/s3_filesys.cc

```diff
@@ -121,6 +136,235 @@ static std::string Sign(const std::string &key,
   return Sign(key, stream.str());
 }
 
+/*!
+ * \brief Generates hash of input as per SHA256 algorithm
+ * Ref: http://stackoverflow.com/questions/2262386/generate-sha256-with-openssl-and-c
+ * \param str input to hash
+ * \param outputBuffer hashed output
+ */
+static void SHA256(const std::string &str, unsigned char *outputBuffer) noexcept {
```

Comment:
how do you assure outputBuffer is the correct size or large enough?

Reply:
Is this a concern even though I call this from one location where I know the size required. Should I also pass size of output buffer and add a check?